Add jackline and (updated) ocaml-dependencies

sternenseemann opened the pull request:
###### Motivation for this change

Addition of `jackline`, a terminal-based XMPP client purely in OCaml.

For `jackline` added missing dependencies:

* `ptime`
* `tls`
* `notty`
* `otr`
* `erm_xmpp_0_3`

and updated the following packages:

* `x509`
* `cstruct`
* `nocrypto`
* `uutf`
* `uuseg`

This PR also incudes minor fixes for:

* ~~~`zarith`~~~ (@vbgl has fixed it on master)
* `ppx_conv_sexp`

###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] OS X
   - [ ] Linux
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

sternenseemann committed:
ocaml-ptime: init at 0.8.2

sternenseemann commented:
Builds:

* [x] `ocamlPackages`
* [x] `ocamlPackages_4_02`
* [ ] `ocamlPackages_4_03`

vbgl commented:
You may want to look at the derivation for `topkg`, and make your `buildPhase` and `installPhase` similar to its: use `-I …` instead of a symlink; move (or link) the result directory to where it is expected by the nixpkgs version of `findlib`.

sternenseemann commented:
@vbgl Thank you, I will have a look at it.

sternenseemann commented:
@vbgl I updated it. That is clearly better.

sternenseemann committed:
ppx_sexp_conv: use sexplib_p4 instead of sexplib

This prevents potential interface incompatibilities if newer compiler versions are used

sternenseemann committed:
ocaml-nocrypto: 0.5.1 -> 0.5.3

sternenseemann committed:
ocaml-cstruct: 1.6.0 -> 2.3.0

This commit also refactors the expression.

sternenseemann committed:
ocaml-x509: 0.5.0 -> 0.5.3

This commit also refactors the expression.

sternenseemann committed:
ocaml-tls: init at 0.7.1

sternenseemann committed:
ocaml-astring: init at 0.8.3

sternenseemann committed:
ocaml-uutf: 0.9.3 -> 0.9.4

This commit also refactors the expression

sternenseemann committed:
ocaml-uuseg: 0.8.0 -> 0.9.0

This commit also refactors the expression.

sternenseemann committed:
ocaml-notty: init at 0.1.1

vbgl commented:
The `installPhase` of `ptime` is not correct yet: you need to move or symlink the `$out/lib/ptime` directory to `$out/lib/ocaml/${ocaml.version}/site-lib`; otherwise, you won’t be able to use that package (at least, `ocamlfind` won’t find it).

sternenseemann commented:
@vbgl Thank you, I fixed it. Is this only an issue if using `opam-installer`?

sternenseemann committed:
ocaml-otr: init at 0.3.3

sternenseemann committed:
ocaml-erm_xmpp_0_3: init at 0.3

sternenseemann committed:
jackline: init at 2016-10-30

sternenseemann commented:
Note: This PR depends on #20086, otherwise `opam` does not build.

Profpatsch commented on pkgs/development/ocaml-modules/x509/default.nix:
```diff
@@ -1,28 +1,30 @@
     description = "X509 (RFC5280) handling in OCaml";
-    platforms = ocaml.meta.platforms or [];
-    license = stdenv.lib.licenses.bsd2;
+    license = stdenv.licenses.bsd2;
```
stdenv.licenses → stdenv.lib.licenses

sternenseemann replied:
Fixed, thank you!

Profpatsch commented on pkgs/applications/networking/instant-messengers/jackline/default.nix:
```diff
@@ -0,0 +1,37 @@
+
+  meta = with stdenv.lib; {
+    homepage = https://github.com/hannesm/jackline;
+    description = "terminal-based XMPP client purely in OCaml";
```
All descriptions start with a capital letter and end with a `.`.

vbgl replied:
[CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md) reads: “`meta.description` should […] Not have a dot at the end”

Profpatsch commented on pkgs/development/ocaml-modules/astring/default.nix:
```diff
@@ -0,0 +1,43 @@
+
+  meta = {
+    homepage = http://erratique.ch/software/ptime;
+    description = "Alternative String module for OCaml";
```
`.`

Profpatsch commented on pkgs/development/ocaml-modules/cstruct/default.nix:
```diff
@@ -1,20 +1,29 @@
-{stdenv, writeText, fetchurl, ocaml, ocplib-endian, sexplib_p4, findlib,
- async_p4 ? null, lwt ? null, camlp4}:
+{stdenv, buildOcaml, fetchFromGitHub, writeText, ocaml, ocplib-endian, sexplib_p4, findlib,
+ async_p4 ? null, lwt ? null, camlp4, ppx_tools ? null}:
```
It’ a lot nicer to explicitly create switches for those, then sane defaults can be enforced directly in the package function:

```nix
{ asyncSupport ? false, async_p4 ? null
, … }:

with stdenv.lib;
assert asyncSupport -> async_p4 != null;
…
  configureFlags = optional asyncSupport async_p4 ++ …
```

sternenseemann replied:
Good point

Profpatsch commented on pkgs/development/ocaml-modules/cstruct/default.nix:
```diff
@@ -1,20 +1,29 @@
+  propagatedBuildInputs = [ocplib-endian sexplib_p4 lwt async_p4 ppx_tools];
+
+  doCheck = true;
+  checkTarget = "test";
```
needed?

sternenseemann replied:
Yes.

Profpatsch replied:
doCheck as well?

Profpatsch commented on pkgs/development/ocaml-modules/erm_xmpp/0.3.nix:
```diff
@@ -0,0 +1,29 @@
+
+  meta = {
+    homepage = https://github.com/hannesm/xmpp;
+    description = "OCaml based XMPP implementation (fork)";
```
.

Profpatsch commented on pkgs/development/ocaml-modules/nocrypto/default.nix:
```diff
@@ -1,28 +1,33 @@
-{ stdenv, fetchzip, ocaml, findlib, cstruct, type_conv, zarith, ounit }:
+{ stdenv, buildOcaml, fetchFromGitHub, ocaml, findlib, cstruct, type_conv, zarith, ounit, ocaml_oasis, ppx_sexp_conv, lwt ? null}:
```
see above cf. `null`

Profpatsch commented on pkgs/development/ocaml-modules/notty/default.nix:
```diff
@@ -0,0 +1,29 @@
+{stdenv, buildOcaml, fetchFromGitHub, findlib, result, uucp, uuseg, uutf, lwt ? null}:
```
`null`

sternenseemann commented on pkgs/development/ocaml-modules/ptime/default.nix:
```diff
@@ -0,0 +1,53 @@
+    # this is a tiny hack to make #use "topfind" work.
+    # the hack used in ocamlPackage.containers didn't work here…
+    ln -s ${mkpath findlib}/topfind topfind
+    ocaml pkg/pkg.ml build --with-js_of_ocaml false
```
Fixed. It is possible with `-I` as @vbgl pointed out earlier.

Profpatsch commented on pkgs/development/ocaml-modules/uutf/default.nix:
```diff
@@ -1,32 +1,38 @@
   webpage = "http://erratique.ch/software/${pname}";
 in
 
-assert stdenv.lib.versionAtLeast ocaml.version "3.12";
```
`minimumSupportedOcamlVersion`

sternenseemann replied:
Is already used here.

Profpatsch commented on pkgs/applications/networking/instant-messengers/jackline/default.nix:
```diff
@@ -0,0 +1,37 @@
+{stdenv, fetchFromGitHub, ocamlPackages}:
+
+assert stdenv.lib.versionAtLeast ocamlPackages.ocaml.version "4.02.2";
```
`minimumSupportedOcamlVersion`

Profpatsch commented on pkgs/development/ocaml-modules/cstruct/default.nix:
```diff
@@ -1,20 +1,29 @@
+{stdenv, buildOcaml, fetchFromGitHub, writeText, ocaml, ocplib-endian, sexplib_p4, findlib,
+ async_p4 ? null, lwt ? null, camlp4, ppx_tools ? null}:
 
 assert stdenv.lib.versionAtLeast (stdenv.lib.getVersion ocaml) "4.01";
```
`minimumSupportedOcamlVersion`

Profpatsch commented on pkgs/development/ocaml-modules/nocrypto/default.nix:
```diff
@@ -1,28 +1,33 @@
-{ stdenv, fetchzip, ocaml, findlib, cstruct, type_conv, zarith, ounit }:
+{ stdenv, buildOcaml, fetchFromGitHub, ocaml, findlib, cstruct, type_conv, zarith, ounit, ocaml_oasis, ppx_sexp_conv, lwt ? null}:
 
 assert stdenv.lib.versionAtLeast ocaml.version "4.01";
```
`minimumSupportedOcamlVersion`

Profpatsch commented on pkgs/development/ocaml-modules/notty/default.nix:
```diff
@@ -0,0 +1,29 @@
+  };
+
+  buildInputs = [ findlib ];
+  propagatedBuildInputs = [ lwt result uucp uuseg uutf ];
```
Aren’t those normal ocamlPackages?

sternenseemann replied:
Could you elaborate a bit?

Profpatsch replied:
Was about the propagated vs normal build inputs, apparently ocaml needs some packages propagated.

sternenseemann replied:
Yeah, indeed. If a package's interface is used in the module, it needs the package propagated, I guess.

Profpatsch commented on pkgs/development/ocaml-modules/notty/default.nix:
```diff
@@ -0,0 +1,29 @@
+
+  meta = with stdenv.lib; {
+    homepage = https://github.com/pqwy/notty/tree/master;
+    description = "Declarative terminal graphics for OCaml";
```
.

Profpatsch commented on pkgs/development/ocaml-modules/otr/default.nix:
```diff
@@ -0,0 +1,43 @@
+    ln -s $out/lib/otr $out/lib/ocaml/${ocaml.version}/site-lib
+  '';
+
+  doCheck = true;
```
needed?

Profpatsch commented on pkgs/development/ocaml-modules/otr/default.nix:
```diff
@@ -0,0 +1,43 @@
+
+  meta = with stdenv.lib; {
+    homepage = https://github.com/hannesm/ocaml-otr;
+    description = "Off-the-record (OTR) messaging protocol, purely in OCaml";
```
.

Profpatsch commented on pkgs/development/ocaml-modules/ptime/default.nix:
```diff
@@ -0,0 +1,47 @@
+
+  buildInputs = [ ocaml findlib ocamlbuild topkg opam ];
+
+  propagatedBuildInputs = [ result ];
```
buildInputs?

sternenseemann replied:
No, is correct like this.

Profpatsch commented on pkgs/development/ocaml-modules/ptime/default.nix:
```diff
@@ -0,0 +1,47 @@
+
+  meta = {
+    homepage = http://erratique.ch/software/ptime;
+    description = "POSIX time for OCaml";
```
.

Profpatsch commented on pkgs/development/ocaml-modules/tls/default.nix:
```diff
@@ -0,0 +1,32 @@
+{stdenv, buildOcaml, fetchFromGitHub, findlib, ocamlbuild, ocaml_oasis, ppx_tools, ppx_sexp_conv, result, x509, nocrypto, cstruct, lwt ? null}:
```
`null`

Profpatsch commented on pkgs/development/ocaml-modules/tls/default.nix:
```diff
@@ -0,0 +1,32 @@
+  configurePhase = "./configure --prefix $out $configureFlags";
+
+  doCheck = true;
+  checkTarget = "test";
```
needed?

sternenseemann replied:
Yes.

Profpatsch commented on pkgs/development/ocaml-modules/tls/default.nix:
```diff
@@ -0,0 +1,32 @@
+
+  meta = with stdenv.lib; {
+    homepage = https://github.com/mirleft/ocaml-tls;
+    description = "TLS in pure OCaml";
```
.

Profpatsch commented on pkgs/development/ocaml-modules/uuseg/default.nix:
```diff
@@ -1,40 +1,39 @@
+  minimumSupportedOcamlVersion = "4.01";
 
-  name = "ocaml-${pname}-${version}";
+  name = "${pname}";
```
name = pname;

Profpatsch commented on pkgs/development/ocaml-modules/uutf/default.nix:
```diff
@@ -1,32 +1,38 @@
 
-assert stdenv.lib.versionAtLeast ocaml.version "3.12";
+buildOcaml rec {
+  name = "${pname}";
```
`name = pname;`

Profpatsch commented on pkgs/development/ocaml-modules/x509/default.nix:
```diff
@@ -1,28 +1,30 @@
   propagatedBuildInputs = [ asn1-combinators nocrypto ];
 
   configureFlags = "--enable-tests";
+  configurePhase = "./configure --prefix $out $configureFlags";
```
Isn’t the prefix set automatically?

sternenseemann replied:
Yes, but as `--prefix=$out` which `./configure` of this package can not parse.

sternenseemann commented:
@vbgl What do you think, do we still have to keep the checks for `lwt`? It seems to build fine with all ocaml versions.

vbgl commented:
Reverting commit  32e86a3 fixes the `core_kernel` build. However, this breaks `nocrypto-0.5.3`. Should we revert this whole PR, which apparently deserves more testing?

Profpatsch commented:
@vbgl yes, let’s do that. OcamlPackages is still in an extremely unstable state.

vbgl commented:
@sternenseemann would you like to open a new PR, or preferably (IMHA) several PRs, so that we can review and test independent changes independently.

Concerning `ptime` two nitpicking remarks: you don’t need to define `ocaml-version` as you can directly refer to `ocaml.version`; the `meta.description` should not end with a dot.

sternenseemann commented:
@vbgl sounds like a plan!
